pkg/utils/grpc.go

```diff
@@ -27,44 +46,136 @@ func NewGrpc(serverType string, port int, certPem, keyPem, rootCertPem []byte) (
-	if err != nil {
-		log.Error().Err(err).Msg("Error setting up mutual tls for GRPC server")
-		return nil, nil, err
+	// #nosec G402: TLS MinVersion too low
+	tlsConfig := tls.Config{
+		InsecureSkipVerify: false,
+		ServerName:         serverType,
+		ClientAuth:         tls.RequireAndVerifyClientCert,
+		MinVersion:         constants.MinTLSVersion,
+		GetConfigForClient: func(*tls.ClientHelloInfo) (*tls.Config, error) {
+			// use lock to prevent concurrent updates and reads to the tls config
+			s.mu.Lock()
+			defer s.mu.Unlock()
+			return &s.config, nil
+		},
 	}
+	mutualTLS := grpc.Creds(credentials.NewTLS(&tlsConfig))
 	grpcOptions = append(grpcOptions, mutualTLS)
 
-	return grpc.NewServer(grpcOptions...), lis, nil
+	s.server = grpc.NewServer(grpcOptions...)
+	return s, lis, nil
+}
+
+// GetServer returns the gRPC server
+func (s *Server) GetServer() *grpc.Server {
+	return s.server
+}
+
+func (s *Server) initConfig() error {
+	grpcCert, err := s.cm.IssueCertificate(
+		s.certName,
+		certificate.Internal,
+		certificate.FullCNProvided())
+	if err != nil {
+		return err
+	}
+
+	certif, err := tls.X509KeyPair(grpcCert.GetCertificateChain(), grpcCert.GetPrivateKey())
+	if err != nil {
+		return fmt.Errorf("failed loading Certificate (%+v) and Key (%+v) PEM files", grpcCert.GetCertificateChain(), grpcCert.GetPrivateKey())
+	}
+
+	certPool := x509.NewCertPool()
+
+	// Load the set of Root CAs
+	if ok := certPool.AppendCertsFromPEM(grpcCert.GetTrustedCAs()); !ok {
+		return fmt.Errorf("failed to append client certs")
+	}
+
+	// use lock to prevent concurrent updates and reads to the tls config
+	s.mu.Lock()
+	// #nosec G402: TLS MinVersion too low
+	s.config = tls.Config{
+		InsecureSkipVerify: false,
+		ServerName:         s.name,
+		ClientAuth:         tls.RequireAndVerifyClientCert,
+		Certificates:       []tls.Certificate{certif},
+		ClientCAs:          certPool,
+		MinVersion:         constants.MinTLSVersion,
+	}
+	s.mu.Unlock()
+
+	return nil
+}
+
+func (s *Server) watchCertRotations(ctx context.Context) error {
+	// listen for certificate rotation first, so we don't miss any events
+	certRotationChan, unsubscribeRotation := s.cm.SubscribeRotations(s.certName)
+	defer unsubscribeRotation()
+
+	// initial initConfig call creates the server certificate
+	if err := s.initConfig(); err != nil {
+		// this is a fatal error on start, we can't continue without a cert
+		return err
+	}
+
+	// Handle the rotations until the context is cancelled
+	go func() {
+		log.Info().Str("grpc", s.name).Str("cn", s.certName).Msg("Listening for certificate rotations")
+		for {
+			select {
+			case <-certRotationChan:
+				log.Debug().Str("grpc", s.name).Str("cn", s.certName).Msg("Certificate rotation was initiated for grpc server")
+				if err := s.initConfig(); err != nil {
+					events.GenericEventRecorder().ErrorEvent(err, events.CertificateIssuanceFailure, "Error rotating the certificate for grpc server")
```

Comment:
Nice to see the consistency here with the webserver rotations.  Don't need to do anything here, but keep an eye on if we can re-use this in someway.  Probably way too early to do this now.